src/probnum/filtsmooth/gaussfiltsmooth/kalman_utils.py

```diff
@@ -0,0 +1,215 @@
+"""Prediction, update, and smoothing implementations for the Kalman filter/smoother."""
+
+import typing
+
+import numpy as np
+
+import probnum.random_variables as pnrv
+
+from .stoppingcriterion import StoppingCriterion
+
+########################################################################################################################
+# Prediction choices
+########################################################################################################################
+
+
+def predict_via_transition(
+    dynamics_model,
+    start,
+    stop,
+    rv,
+    _intermediate_step=None,
+    _linearise_at=None,
+    _diffusion=1.0,
+) -> (pnrv.RandomVariable, typing.Dict):
+    """Compute the prediction under the assumption that the transition is available in
+    closed form."""
+    return dynamics_model.transition_rv(
+        rv=rv,
+        start=start,
+        stop=stop,
+        step=_intermediate_step,
+        _linearise_at=_linearise_at,
+        _diffusion=_diffusion,
+    )
+
+
+########################################################################################################################
+# Measure choices
+########################################################################################################################
+
+
+def measure_via_transition(
+    measurement_model,
+    rv,
+    time,
+    _linearise_at=None,
+) -> (pnrv.RandomVariable, typing.Dict):
+    """Compute the measurement under the assumption that the transition is available in
+    closed form."""
+    return measurement_model.transition_rv(
+        rv=rv, start=time, _linearise_at=_linearise_at
+    )
+
+
+########################################################################################################################
+# Update choices
+########################################################################################################################
+
+
+def update_classic(
+    measurement_model, rv, time, data, _linearise_at=None
+) -> (pnrv.RandomVariable, typing.Dict):
+    """Compute a classic Kalman update.
+
+    This is a combination of a call to measure_*() and a Gaussian
+    update.
+    """
+    meas_rv, info = measure_via_transition(
+        measurement_model=measurement_model,
+        rv=rv,
+        time=time,
+        _linearise_at=_linearise_at,
+    )
+    crosscov = info["crosscov"]
+    updated_rv = condition_state_on_measurement(
+        pred_rv=rv, meas_rv=meas_rv, crosscov=crosscov, data=data
+    )
+    return updated_rv, meas_rv, {}
+
```

Comment:
here, the joseph updates and square-root updates can be added. As long as they have the same signature, they are immediately usable inside Kalman.